File: packages/protocol-ts/src/validation.ts
/**
 * Validation for WrapDrive protocol messages.
 *
 * Validators return a {@link ValidationResult} rather than throwing, so callers
 * (HTTP handlers, the negotiator) can decide how to surface failures. Each
 * failure identifies the offending field, satisfying the protocol requirement
 * that rejections name the invalid field.
 */

import { APP_PROTOCOL, MAX_ALIAS_LENGTH, MAX_PORT, MIN_PORT } from './constants.js';
import type { Capabilities, DeviceInfo, Protocol } from './models.js';

/** A single validation failure, naming the field and why it was rejected. */
export interface ValidationError {
  /** Dotted path of the offending field, e.g. `"port"` or `"minChunkSize"`. */
  field: string;
  /** Human-readable explanation of the violation. */
  message: string;
}

/** The outcome of validating a message: either valid, or a list of errors. */
export type ValidationResult = { valid: true } | { valid: false; errors: ValidationError[] };

const VALID_PROTOCOLS: readonly Protocol[] = ['http', 'https'];

const VALID = (): ValidationResult => ({ valid: true });

const INVALID = (errors: ValidationError[]): ValidationResult => ({
  valid: false,
  errors,
});

function isInteger(value: number): boolean {
  return Number.isInteger(value);
}

/**
 * Validate a {@link DeviceInfo} against the protocol rules:
 * non-empty alias <= 64 chars, port in [1024, 65535], non-empty fingerprint,
 * and a `protocol` of `http` or `https`.
 */
export function validateDeviceInfo(info: DeviceInfo): ValidationResult {
  const errors: ValidationError[] = [];

  if (info.alias.length === 0) {
    errors.push({ field: 'alias', message: 'alias must not be empty' });
  } else if (info.alias.length > MAX_ALIAS_LENGTH) {
    errors.push({
      field: 'alias',
      message: `alias must be at most ${MAX_ALIAS_LENGTH} characters`,
    });
  }

  if (!isInteger(info.port) || info.port < MIN_PORT || info.port > MAX_PORT) {
    errors.push({
      field: 'port',
      message: `port must be an integer within [${MIN_PORT}, ${MAX_PORT}]`,
    });
  }

  if (info.fingerprint.length === 0) {
    errors.push({ field: 'fingerprint', message: 'fingerprint must not be empty' });
  }

  if (!VALID_PROTOCOLS.includes(info.protocol)) {
    errors.push({ field: 'protocol', message: "protocol must be 'http' or 'https'" });
  }

  return errors.length === 0 ? VALID() : INVALID(errors);
}

/**
 * Validate a {@link Capabilities} against the protocol rules:
 * `minChunkSize > 0`, `minChunkSize <= maxChunkSize`, and
 * `maxParallelConnections >= 1`. The `appProtocol` is also checked to be the
 * WrapDrive identifier.
 */
export function validateCapabilities(caps: Capabilities): ValidationResult {
  const errors: ValidationError[] = [];

  if (caps.appProtocol !== APP_PROTOCOL) {
    errors.push({
      field: 'appProtocol',
      message: `appProtocol must be '${APP_PROTOCOL}'`,
    });
  }

  if (!isInteger(caps.minChunkSize) || caps.minChunkSize <= 0) {
    errors.push({
      field: 'minChunkSize',
      message: 'minChunkSize must be a positive integer',
    });
  }

  if (!isInteger(caps.maxChunkSize) || caps.maxChunkSize <= 0) {
    errors.push({
      field: 'maxChunkSize',
      message: 'maxChunkSize must be a positive integer',
    });
  }

  if (
    isInteger(caps.minChunkSize) &&
    isInteger(caps.maxChunkSize) &&
    caps.minChunkSize > caps.maxChunkSize
  ) {
    errors.push({
      field: 'minChunkSize',
      message: 'minChunkSize must not exceed maxChunkSize',
    });
  }

  if (!isInteger(caps.maxParallelConnections) || caps.maxParallelConnections < 1) {
    errors.push({
      field: 'maxParallelConnections',
      message: 'maxParallelConnections must be an integer >= 1',
    });
  }

  return errors.length === 0 ? VALID() : INVALID(errors);
}

/** Convenience guard: `true` when the {@link DeviceInfo} passes validation. */
export function isValidDeviceInfo(info: DeviceInfo): boolean {
  return validateDeviceInfo(info).valid;
}

/** Convenience guard: `true` when the {@link Capabilities} passes validation. */
export function isValidCapabilities(caps: Capabilities): boolean {
  return validateCapabilities(caps).valid;
}

File: apps/android/core/protocol/src/main/kotlin/com/wrapdrive/core/protocol/Validation.kt
package com.wrapdrive.core.protocol

/** A single validation failure, naming the field and the reason. */
data class ValidationError(val field: String, val message: String)

/** The outcome of validating a message. */
sealed interface ValidationResult {
    val isValid: Boolean

    data object Valid : ValidationResult {
        override val isValid: Boolean = true
    }

    data class Invalid(val errors: List<ValidationError>) : ValidationResult {
        override val isValid: Boolean = false
    }
}

/**
 * Validation for WrapDrive protocol messages (Kotlin port).
 *
 * Mirrors the TypeScript validators: non-empty alias <= 64 chars, port in
 * [1024, 65535], non-empty fingerprint, valid capability bounds. Each failure
 * names the offending field.
 */
object Validation {
    fun validateDeviceInfo(info: DeviceInfo): ValidationResult {
        val errors = mutableListOf<ValidationError>()

        if (info.alias.isEmpty()) {
            errors += ValidationError("alias", "alias must not be empty")
        } else if (info.alias.length > WrapDriveProtocol.MAX_ALIAS_LENGTH) {
            errors +=
                ValidationError(
                    "alias",
                    "alias must be at most ${WrapDriveProtocol.MAX_ALIAS_LENGTH} characters",
                )
        }

        if (info.port < WrapDriveProtocol.MIN_PORT || info.port > WrapDriveProtocol.MAX_PORT) {
            errors +=
                ValidationError(
                    "port",
                    "port must be within [${WrapDriveProtocol.MIN_PORT}, ${WrapDriveProtocol.MAX_PORT}]",
                )
        }

        if (info.fingerprint.isEmpty()) {
            errors += ValidationError("fingerprint", "fingerprint must not be empty")
        }

        return if (errors.isEmpty()) ValidationResult.Valid else ValidationResult.Invalid(errors)
    }

    fun validateCapabilities(caps: Capabilities): ValidationResult {
        val errors = mutableListOf<ValidationError>()

        if (caps.appProtocol != WrapDriveProtocol.APP_PROTOCOL) {
            errors +=
                ValidationError("appProtocol", "appProtocol must be '${WrapDriveProtocol.APP_PROTOCOL}'")
        }

        if (caps.minChunkSize <= 0) {
            errors += ValidationError("minChunkSize", "minChunkSize must be a positive integer")
        }

        if (caps.maxChunkSize <= 0) {
            errors += ValidationError("maxChunkSize", "maxChunkSize must be a positive integer")
        }

        if (caps.minChunkSize > 0 && caps.maxChunkSize > 0 && caps.minChunkSize > caps.maxChunkSize) {
            errors += ValidationError("minChunkSize", "minChunkSize must not exceed maxChunkSize")
        }

        if (caps.maxParallelConnections < 1) {
            errors +=
                ValidationError("maxParallelConnections", "maxParallelConnections must be >= 1")
        }

        return if (errors.isEmpty()) ValidationResult.Valid else ValidationResult.Invalid(errors)
    }
}
